Add Sentry integration for Backend and Frontend

richard015ar opened the pull request:
Related issue: https://github.com/pressbooks/private/issues/374

This change implements a new Sentry Integration to PB, and also add a JavaScript observer to monitor performance and catch front end issues.
Details about how it works and documentation: https://docs.google.com/document/d/1hg35gRWlxcQqC2DElrBi2btLFQQXvMgs-NhgvmnMqpA/edit

### How to test
- Create and set the Environment variables following [docs](https://docs.google.com/document/d/1hg35gRWlxcQqC2DElrBi2btLFQQXvMgs-NhgvmnMqpA/edit) in your local environment (make sure `ENABLE_SENTRY=1`
- Force any error or exception in any enable plugin or theme within your code
- Check [Sentry pressbooks project](https://sentry.io/organizations/pressbooks/issues) issues and make sure the issue is there.

_Note there are new ENV variables for sentry and also we can  delete the previous one in Staging after apply this change._

richard015ar commented:
A change was added: `twentyseventeen`  by `twentytwenty`  in test_allowedRootThemes. Looks like twentyseventeen isn’t found as is being filtered out of the input array, seems twentyseventeen is not bundled with newer WordPress versions.
Related issue: https://github.com/pressbooks/pressbooks/issues/2064

I will add more unit tests to cover >= 0.02% required.

ho-man-chan commented on assets/src/scripts/sentry.js:
```diff
@@ -0,0 +1,18 @@
+Sentry.init( {
+	dsn: SentryParams.dsn,
+	integrations: [ new Integrations.BrowserTracing() ],
+	tracesSampleRate: 1.0,
```
curious how the sampling rate will affect server performance. Will observe during testing in staging server.

richard015ar replied:
I suggest do it dynamically, so it can be configured different for each server. An new ENV variable could solve it, since it will depend on server hardware, bandwidth and other resources.

ho-man-chan commented on hooks.php:
```diff
@@ -346,11 +346,8 @@
-	! is_null( env( 'SENTRY_KEY' ) ) &&
-	! is_null( env( 'SENTRY_ORGANIZATION' ) ) &&
-	! is_null( env( 'SENTRY_PROJECT' ) )
+	! is_null( env( 'ENABLE_SENTRY' ) ) &&
```
Concern for backwards compatibility since environment variables are removed and new ones are added.

richard015ar replied:
It is compatible, since other variables (the previous ones) are not longer used, then are not considered in any part of the code.

ho-man-chan commented on inc/class-sentry.php:
```diff
@@ -0,0 +1,159 @@
+	 */
+	public function getCurrentUserForTracking() {
+		$user_id = get_current_user_id();
+		$this->user = $user_id > 0 ? $user = get_userdata( $user_id ) : false;
```
may reduce to 
`$this->user = $user_id > 0 ? get_userdata( $user_id ) : false;`

richard015ar replied:
Yes! I will change it. It was an typing error, doesn't make sense how I did it, probably a typing error. Thank you for catching it!

ho-man-chan commented on inc/class-sentry.php:
```diff
@@ -0,0 +1,159 @@
+					'email' => $this->user->user_email,
+				];
+			}
+			wp_localize_script( self::WP_SCRIPT_NAME, 'SentryParams', $script_params );
```
What's the purpose of this?

richard015ar replied:
It is used for pass variables to the JavaScript script from server side before create and compile the script. More references here:
- https://wpshout.com/building-a-magical-golden-bridge-from-php-to-javascript-with-wp_localize_script/
- https://medium.com/vunamhung/passing-php-variables-to-javascript-with-wp-localize-script-d99e165c1f9c

ho-man-chan commented on inc/utility/namespace.php:
```diff
@@ -1500,22 +1500,3 @@ function do_shortcode_by_tags( $content, array $tags ) {
 	return $shortcoded;
 }
-
-function initialize_sentry() {
```
Is this possible to reuse also the existing sentry environment variables so it's backward compatible?

richard015ar replied:
yes, is possible. I can add that compatibility if it is convenient for you. Probably I would need document the reasons for future references. Could you help me with that?

ho-man-chan commented on tests/test-sentry.php:
```diff
@@ -1,22 +1,68 @@
+	/**
+	 * Test init Sentry function to test instance returned.
+	 */
+	public function test_getInstance() {
```
Additional negative test cases would increase the robustness of the feature.
Example:
1. Missing environment variable -> application still functioning as expected
2. No user signed in -> application still logs with user as null

Test to ensure the errors are actually tracked by Sentry would be a great validation of the feature working with simulated errors.

richard015ar replied:
Makes sense! I will add those.

ho-man-chan reviewed (commented):
Great use of OOP Concept! 👍 
I notice some concerns for backwards compatibility on the use of environment variable. It would be perfect if the previous environment variable can work with new features. This will greatly reduce the headache of Open source users.

Furthermore, I suggest to add testing negative cases (ex: missing environment variables or non logged user) to make sure the feature would still work and not explode.

Finally, a test to ensure simulated errors are actually sent to sentry would be perfect if possible...

ho-man-chan commented:
Appreciated. 🥇

richard015ar commented on assets/src/scripts/sentry.js:
```diff
@@ -0,0 +1,18 @@
+Sentry.init( {
+	dsn: SentryParams.dsn,
+	integrations: [ new Integrations.BrowserTracing() ],
+	tracesSampleRate: SentryParams.sample,
```
Sample rate variable

richard015ar commented on inc/class-pressbookssentry.php:
```diff
@@ -0,0 +1,224 @@
+
+	const DEFAULT_ENVIRNOMENT = 'staging';
+
+	const DEFAULT_TRACE_SAMPLE_RATE = 0.5;
```
Default value for sample rate

richard015ar commented on inc/class-pressbookssentry.php:
```diff
@@ -0,0 +1,224 @@
+
+	const ENABLED_VALUE = 1;
+
+	const ENVIRONMENT_VARIABLES_V1 = 1;
```
See https://docs.google.com/document/d/1hg35gRWlxcQqC2DElrBi2btLFQQXvMgs-NhgvmnMqpA/edit#

richard015ar commented on inc/class-pressbookssentry.php:
```diff
@@ -0,0 +1,224 @@
+			self::$instance->setSentryDSNFromEnvironmentVariables();
+			self::$instance->getCurrentUserForTracking();
+			if (
+				self::environmentVariablesType() === self::ENVIRONMENT_VARIABLES_V1 ||
```
If [previous integration version](https://github.com/pressbooks/private/issues/130#issuecomment-634938654) 
```
SENTRY_KEY=
SENTRY_ORGANIZATION=
SENTRY_PROJECT=
```
is implemented, only PHP issues will be captured by Sentry.

richard015ar commented on inc/class-pressbookssentry.php:
```diff
@@ -0,0 +1,224 @@
+	 */
+	static public function environmentVariablesType() {
+		if ( self::areEnvironmentVariablesPresent() ) {
+			return (
```
See: https://github.com/pressbooks/private/issues/130#issuecomment-634938654

richard015ar commented on tests/test-pressbooks.php:
```diff
@@ -33,10 +33,10 @@ public function test_allowedBookThemes() {
 	 */
 	public function test_allowedRootThemes() {
-		$result = $this->pb->allowedRootThemes( [ 'pressbooks-book' => true, 'pressbooks-clarke' => true, 'pressbooks-fake' => true, 'twentyseventeen' => true ] );
+		$result = $this->pb->allowedRootThemes( [ 'pressbooks-book' => true, 'pressbooks-clarke' => true, 'pressbooks-fake' => true, 'twentytwenty' => true ] );
```
Fix for: https://github.com/pressbooks/pressbooks/issues/2064

richard015ar commented:
@ho-man-chan, I implemented some changes:
- I added environment variables compatibility with [the previous ones](https://github.com/pressbooks/private/issues/130#issuecomment-634938654)
- I added more use cases considering negative cases and with different environment variables cases.
- I added `SENTRY_TRACE_SAMPLE_RATE` to be able to manipulate the sample rate using env variable. By default the value is 0.5.
- I updated the documentation: https://docs.google.com/document/d/1hg35gRWlxcQqC2DElrBi2btLFQQXvMgs-NhgvmnMqpA/edit#

ho-man-chan reviewed (approved):
Looks good!! Nice tests!
